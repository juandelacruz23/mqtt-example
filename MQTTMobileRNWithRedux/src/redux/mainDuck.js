/**
 * What is a duck?
 * https://github.com/erikras/ducks-modular-redux
 */

import { connectionStatuses, subscriptionStatuses } from "../statuses";

const CHANGE_AND_PUSH = "CHANGE_AND_PUSH";
const CHANGE_VALUE = "CHANGE_VALUE";
const PUSH_TEXT = "PUSH_TEXT";

export const changeAndPush = (newValue, textToPush) => ({
  type: CHANGE_AND_PUSH,
  payload: {
    value: newValue,
    text: textToPush,
  },
});

export const changeValue = newValue => ({
  type: CHANGE_VALUE,
  payload: newValue,
});

export const pushText = text => ({
  type: PUSH_TEXT,
  payload: text,
});

const initialState = {
  text: [],
  connectionStatus: connectionStatuses.DISCONNECTED,
  subscriptionStatus: subscriptionStatuses.UNSUBSCRIBED,
  host: "",
  port: "",
  topic: "",
  showMessageDialog: false,
  message: "",
  connecting: false,
};

export function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case CHANGE_AND_PUSH:
      return {
        ...state,
        ...action.payload.value,
        text: [...state.text, action.payload.text],
      };
    case CHANGE_VALUE:
      return { ...state, ...action.payload };
    case PUSH_TEXT:
      return { ...state, text: [...state.text, action.payload] };
    default:
      return state;
  }
}
