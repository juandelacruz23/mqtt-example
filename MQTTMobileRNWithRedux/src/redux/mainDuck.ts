/**
 * What is a duck?
 * https://github.com/erikras/ducks-modular-redux
 */

import { Action } from "redux";
import ConnectionStatus from "../ConnectionStatus";
import SubscriptionStatus from "../SubscriptionStatus";

const CHANGE_AND_PUSH = "CHANGE_AND_PUSH";
const CHANGE_VALUE = "CHANGE_VALUE";
const PUSH_TEXT = "PUSH_TEXT";

export interface AppAction extends Action<string> {
  type: string,
  payload?: any,
}

export const changeAndPush = (newValue: Object, textToPush: string) : AppAction => ({
  type: CHANGE_AND_PUSH,
  payload: {
    value: newValue,
    text: textToPush,
  },
});

export const changeValue = (newValue: Object) : AppAction => ({
  type: CHANGE_VALUE,
  payload: newValue,
});

export const pushText = (text: string): AppAction => ({
  type: PUSH_TEXT,
  payload: text,
});

export interface AppState {
  connectionStatus: ConnectionStatus,
  host: string,
  loading: boolean,
  message: string,
  port: string,
  showMessageDialog: boolean,
  subscriptionStatus: SubscriptionStatus,
  text: string[],
  topic: string,
}

const initialState: AppState = {
  text: [],
  connectionStatus: ConnectionStatus.DISCONNECTED,
  subscriptionStatus: SubscriptionStatus.UNSUBSCRIBED,
  host: "",
  port: "",
  topic: "",
  showMessageDialog: false,
  message: "",
  loading: false,
};

export function reducer(state = initialState, action: AppAction = { type: 'DEFAULT'} ): AppState {
  switch (action.type) {
    case CHANGE_AND_PUSH:
      return {
        ...state,
        ...action.payload.value,
        text: [...state.text, action.payload.text],
      };
    case CHANGE_VALUE:
      return { ...state, ...action.payload };
    case PUSH_TEXT:
      return { ...state, text: [...state.text, action.payload] };
    default:
      return state;
  }
}
