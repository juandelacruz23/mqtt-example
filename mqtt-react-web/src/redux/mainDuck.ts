import { Action } from "redux";
import MQTTOptions from "../paho.mqtt.types/MQTTOptions";

const CHANGE_VALUE = "CHANGE_VALUE";
export const CONNECT = "CONNECT";
export const DISCONNECT = "DISCONNECT";
export const CONSOLE_EVENT = "CONSOLE_EVENT";

export interface AppAction<T> extends Action<string> {
  type: string;
  payload?: T;
}

export type FixedAppAction<T> = Required<AppAction<T>>;

export type BaseAction = FixedAppAction<object>;

export function changeValue(newValue: object): BaseAction {
  return {
    type: CHANGE_VALUE,
    payload: newValue,
  } as const;
}

export function connectClient(
  mqttOptions: MQTTOptions,
): FixedAppAction<MQTTOptions> {
  return {
    type: CONNECT,
    payload: mqttOptions,
  } as const;
}

export type MQTTConnectAction = ReturnType<typeof connectClient>;

export function consoleEvent(
  event: ConsoleEvent,
): FixedAppAction<ConsoleEvent> {
  return {
    type: CONSOLE_EVENT,
    payload: event,
  } as const;
}

export type ConsoleEventAction = ReturnType<typeof consoleEvent>;

export function disconnectClient(): AppAction<undefined> {
  return {
    type: DISCONNECT,
  };
}

const INITIAL_STATE = {
  host: "",
  isConnected: false,
  port: 0,
  clientId: ``,
  path: "",
  events: [
    {
      value: "INFO - Starting Eclipse Paho JavaScript Utility.",
      timestamp: Date.now(),
    },
  ],
};

export type AppState = typeof INITIAL_STATE;

export function reducer(
  state: AppState = INITIAL_STATE,
  action: AppAction<object> = { type: "DEFAULT" },
): AppState {
  switch (action.type) {
    case CHANGE_VALUE:
    case CONNECT:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
}

export default {
  connectClient,
  changeValue,
  disconnectClient,
};
