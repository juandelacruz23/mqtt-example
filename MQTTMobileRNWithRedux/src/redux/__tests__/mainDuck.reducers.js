import { reducer, changeValue, pushText } from "../mainDuck";

describe("main reducer", () => {
  it("should return initialState as default", () => {
    expect(reducer()).toMatchSnapshot();
  });

  it("should handle CHANGE_VALUE action with initialState", () => {
    const action1 = changeValue({ host: "fake value" });
    expect(reducer(undefined, action1)).toHaveProperty("host", "fake value");
  });

  it("should handle PUSH_TEXT action with initialState", () => {
    const { text: result1 } = reducer(undefined, pushText("fake text 1"));
    const { text: result2 } = reducer(undefined, pushText("fake text 2"));
    expect(result1).toContain("fake text 1");
    expect(result2).toContain("fake text 2");
  });
});
