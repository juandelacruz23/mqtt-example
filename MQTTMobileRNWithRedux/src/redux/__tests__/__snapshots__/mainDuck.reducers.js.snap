// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main reducer should return initialState as default 1`] = `
Object {
  "connectionStatus": 1,
  "host": "",
  "loading": false,
  "message": "",
  "port": "",
  "showMessageDialog": false,
  "subscriptionStatus": 1,
  "text": Array [],
  "topic": "",
}
`;
